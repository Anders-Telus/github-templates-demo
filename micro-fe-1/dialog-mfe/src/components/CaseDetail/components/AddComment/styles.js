import styled from 'styled-components'

export const AddCommentWrapper = styled.div`
  padding: 1rem;
  overflow: hidden;
  padding: ${props => (props.type === 'case' ? '1rem 1rem 1rem 0' : '1rem')};
`

export const AddCommentSection = styled.div`
  float:left;
  min-height:36px;
  width: ${props => (props.customWidth)};
`
export const CommentsButtonSection = styled.div`
  width:${props => (props.buttonWidth)};
  min-height:36px;
  display: flex;
  justify-content: flex-start;
  height: 100%;
  left: ${props => (props.left)};
`

export const SaveButtonWrapper = styled.div`
  bottom: 32px;
  left: 0px;
  margin-left: 14px;
`

export const Editable = styled.div`
  width: 100%;
  outline: none;
  margin-top:5px;
  overflow-x: hidden;
  &:focus {
    outline: none;
  }

  &::empty:before {
    content: attr(placeholder);
    color: #555; 
  }

  :empty:focus:before {
    content: "";
}
`
