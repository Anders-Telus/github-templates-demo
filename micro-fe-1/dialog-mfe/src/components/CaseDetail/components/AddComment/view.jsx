/* eslint-disable jsx-a11y/aria-role */
import React, { useState, useEffect, useRef } from 'react'
import PropTypes from 'prop-types'
import Spinner from '@tds/core-spinner'
import momentTZ from 'moment-timezone'
import {
  AddCommentWrapper,
  AddCommentSection,
  CommentsButtonSection,
  SaveButtonWrapper,
  Editable
} from './styles'
import { getCasaAppLocale } from '../../../../../../utils/helper'
import {
  singleQuotes,
  singleQuotesCaseComment
} from '../../../../../../constant'
import {
  SaveButton
} from '../../styles'

const AddComment = ({
  updateComment,
  agentIdentity,
  tabUniqId,
  taskId,
  billingAcctNum,
  type,
  taskMetaData,
  addCommentStatus,
  updateCommentStatus,
  newCommentText,
  updateCommentText,
  handleCommentText,
  saveButtonActive,
  textAreaActive,
  handleSpinner,
  traceId,
  caseId,
  updateTracker,
  trackUpdateInFields,
  disabled
}) => {
  const locale = getCasaAppLocale()
  const currentLocaleLang = momentTZ.locale()
  const [divStyle, setDivStyle] = useState({
    background: 'transparent',
    width: '100%',
    minHeight: '30px',
    maxHeight: '150px',
    overflowY: 'scroll',
    borderBottom: '2px solid #d8d8d8'
  })
  const [cellValue, setCellValue] = useState(newCommentText || '')
  const [newComment, setNewComment] = useState(newCommentText)
  const [activeSaveButton, setActiveSaveButton] = useState(saveButtonActive)
  const [activeTextArea, setActiveTextArea] = useState(textAreaActive)
  const cellRef = useRef()
  const status = addCommentStatus === 'PENDING'
  const formatComment = comment => comment.trim().replace(/[^\S\r\n]+/g, ' ').replace(/'/g, singleQuotes)
  const formatCommentCase = comment => comment.trim().replace(/[^\S\r\n]+/g, ' ').replace(/'/g, singleQuotesCaseComment)
  const saveComment = () => {
    if (taskId) {
      const lastUpdatedBy = { id: agentIdentity.employeeId, name: `${agentIdentity.firstName} ${agentIdentity.lastName}` }
      const body = {
        event: {
          StandAloneTask: {
            taskId,
            lastUpdatedBy,
            comments: [
              {
                createdAt: (new Date()).toISOString(),
                createdBy: 'T-989898',
                text: formatComment(newCommentText)
              }
            ],
            accountRef: {
              accountNumber: billingAcctNum,
              systemId: ''
            }
          }
        },
        taskType: type,
        eventType: 'UITaskUpdate',
        externalId: taskMetaData && taskMetaData.externalId,
        systemSourceId: 'CasaUI',
        timeOccurred: ((new Date()).toISOString()),
        traceId: taskMetaData && taskMetaData.interactionId,
        lob: taskMetaData && taskMetaData.lob,
        brand: taskMetaData && taskMetaData.brand
      }
      handleSpinner()
      updateComment(body, tabUniqId, taskId)
    } else {
      // case comment update
      const body = {
        traceId,
        relatedEntity: [
          {
            tickets: {
              caseComments: JSON.stringify({ slug: '', text: formatCommentCase(newCommentText) }),
              lastUpdatedBy: { id: agentIdentity.employeeId, name: `${agentIdentity.firstName} ${agentIdentity.lastName}` }
            }
          }
        ]
      }
      handleSpinner()
      updateComment(body, tabUniqId, caseId)
    }
  }
  const onCl = () => {
    if (!disabled) {
      setDivStyle(p => ({
        ...p, background: 'white', borderBottom: '2px solid #4b286d', outline: 'none'
      }))
    }
  }
  const onBl = () => {
    const { textContent } = cellRef.current
    if (!textContent) {
      setDivStyle(p => ({
        ...p, background: 'transparent', borderBottom: '2px solid #d8d8d8', outline: 'none'
      }))
    }
  }
  const pushFieldsUpdate = (value) => {
    const commentType = type === 'case' ? 'caseComments' : 'taskComments'
    if (updateTracker[commentType] !== value) {
      trackUpdateInFields(tabUniqId, commentType, value)
    }
  }
  const onInput = (e) => {
    const { textContent } = e.currentTarget
    // only for spaces
    if (textContent.length) {
      setActiveTextArea(true)
    }
    if (textContent.trim() !== '') {
      if ((/\w|\W/g).test(textContent)) {
        setActiveSaveButton(true)
        pushFieldsUpdate(false)
        setActiveTextArea(true)
        const formState = {
          saveButtonActive: true,
          textAreaActive: true,
          newCommentText: cellRef.current.innerText
        }
        setDivStyle(p => ({
          ...p, background: 'white', borderBottom: '2px solid #4b286d', outline: 'none'
        }))
        handleCommentText(taskId, formState)
      } else {
        setActiveSaveButton(false)
        pushFieldsUpdate(true)
      }
    } else if (textContent.length) {
      setActiveSaveButton(false)
      pushFieldsUpdate(true)
      const state = {
        saveButtonActive: false,
        textAreaActive: true,
        caseCommentText: textContent.trim()
      }
      handleCommentText(taskId, state)
    } else {
      cellRef.current.innerHTML = ''
      setCellValue('')
      if (textContent.length === 0) {
        cellRef.current.blur()
        setDivStyle(p => ({
          ...p, background: 'transparent', borderBottom: '2px solid #d8d8d8', outline: 'none'
        }))
        setActiveSaveButton(false)
        pushFieldsUpdate(true)
      } else {
        setDivStyle(p => ({
          ...p, background: 'white', borderBottom: '2px solid #4b286d', outline: 'none'
        }))
        setActiveSaveButton(true)
        pushFieldsUpdate(false)
      }
      const formState = {
        saveButtonActive: false,
        textAreaActive: false,
        newCommentText: textContent
      }
      handleCommentText(taskId, formState)
    }
  }
  const textareaId = type === 'case' ? `textarea-${type}` : 'textarea-task'
  const commentSaveButtonId = type === 'case' ? 'comment_button_case' : 'comment_button_task'
  useEffect(() => {
    if (addCommentStatus === 'SUCCESS') {
      setNewComment('')
      setActiveSaveButton(false)
      pushFieldsUpdate(true)
      setActiveTextArea(false)
      const formState = {
        saveButtonActive: false,
        textAreaActive: false,
        newCommentText: ''
      }
      setDivStyle(p => ({
        ...p, background: 'transparent', borderBottom: '2px solid #d8d8d8', outline: 'none'
      }))
      setCellValue('')
      handleCommentText(taskId, formState)
      updateCommentStatus(tabUniqId, 'UNINIT', taskId)
      cellRef.current.innerText = ''
    }
    if (!activeTextArea) {
      setDivStyle(p => ({
        ...p, background: 'transparent', borderBottom: '2px solid #d8d8d8', outline: 'none'
      }))
    }
    if (cellValue) {
      setDivStyle(p => ({
        ...p, background: 'white', borderBottom: '2px solid #4b286d', outline: 'none'
      }))
    }
    // @todo this is not required
    updateCommentText(tabUniqId, newComment)
  }, [addCommentStatus, newComment])
  let width = ''
  let left = ''
  let buttonWidth = ''
  if (type === 'case') {
    if (currentLocaleLang === 'fr') {
      width = '480px'
      buttonWidth = '106px'
      left = '498px'
    } else {
      width = '500px'
      buttonWidth = '66px'
      left = '516px'
    }
  } else if (currentLocaleLang === 'fr') {
    width = '430px'
    buttonWidth = '106px'
    left = '460px'
  } else {
    width = '471px'
    buttonWidth = '66px'
    left = '502px'
  }

  return (
    <AddCommentWrapper type={type}>
      <AddCommentSection customWidth={width}>
        <Editable
          className="editable"
          ref={cellRef}
          role="comment_textarea"
          id={textareaId}
          contentEditable={!disabled}
          style={divStyle}
          onClick={onCl}
          onBlur={onBl}
          onInput={onInput}
          content={cellValue.length}
          placeholder={locale.case.addComment}
          active={activeTextArea}
          maxLength="2000"
          dangerouslySetInnerHTML={{ __html: cellValue.replace(/\n/g, '<br/>') }}
        />
      </AddCommentSection>
      <CommentsButtonSection left={left} width={buttonWidth}>
        <SaveButtonWrapper>
          <Spinner
            spinning={status}
            size="small"
            label={locale.task.form.addCommentSpinnerLabel}
            inline
          >
            <SaveButton
              // active={activeSaveButton}
              onClick={() => saveComment(taskId)}
              id={commentSaveButtonId}
              disabled={!activeSaveButton}
              // cursorChange={disabled}
            >
              {locale.case.buttons.save}
            </SaveButton>
          </Spinner>
        </SaveButtonWrapper>
      </CommentsButtonSection>
    </AddCommentWrapper>
  )
}
AddComment.defaultProps = {
  newCommentText: '',
  traceId: '',
  caseId: null,
  disabled: false
}
AddComment.propTypes = {
  updateComment: PropTypes.func.isRequired,
  agentIdentity: PropTypes.object.isRequired,
  tabUniqId: PropTypes.string.isRequired,
  taskId: PropTypes.string.isRequired,
  addCommentStatus: PropTypes.string.isRequired,
  updateCommentStatus: PropTypes.func.isRequired,
  newCommentText: PropTypes.string,
  type: PropTypes.string.isRequired,
  updateCommentText: PropTypes.func.isRequired,
  handleCommentText: PropTypes.func.isRequired,
  saveButtonActive: PropTypes.bool.isRequired,
  textAreaActive: PropTypes.bool.isRequired,
  handleSpinner: PropTypes.func.isRequired,
  billingAcctNum: PropTypes.string.isRequired,
  taskMetaData: PropTypes.object.isRequired,
  traceId: PropTypes.string,
  caseId: PropTypes.number,
  updateTracker: PropTypes.object.isRequired,
  trackUpdateInFields: PropTypes.func.isRequired,
  disabled: PropTypes.bool
}
export default AddComment
