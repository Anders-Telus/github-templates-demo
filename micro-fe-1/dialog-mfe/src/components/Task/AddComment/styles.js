import styled from 'styled-components'
import {
  colorCardinal
} from '@tds/core-colours'

export const TextEditable = styled.div`
  font-weight: 500;
  font-style: normal;
  font-size: 14px;
  letter-spacing: -0.6px;
  width: 563px !important;
  outline: none;
  &:focus {
    outline: none;
  }

  &:empty:before {
    content: attr(placeholder);
    color: ${colorCardinal} !important;
    opacity: 0.85;
  }

  :empty:focus:before {
    content: "";
  }
  & > div {
    letter-spacing: -0.6px;
  }
`
export const TextEditableWrapper = styled.div`
 width: auto;
`
