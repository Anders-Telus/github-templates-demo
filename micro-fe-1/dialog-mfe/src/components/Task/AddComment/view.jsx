import React, { useState, useRef } from 'react'
import PropTypes from 'prop-types'
import {
  TextEditable
} from './styles'


const SelectInput = ({
  id, placeholder, addCommentHandler, newCommentText, disabled
}) => {
  const cellRef = useRef()
  const [cellValue, setCellValue] = useState(newCommentText || '')
  const [divStyle, setDivStyle] = useState({
    background: 'transparent',
    width: '100%',
    minHeight: '30px',
    maxHeight: '132px',
    overflowY: 'scroll',
    borderBottom: '2px solid #d8d8d8',
    padding: '9px 8px 9px 8px'
  })

  const editableClick = () => {
    if (!disabled) {
      setDivStyle(p => ({
        ...p, background: 'white', borderBottom: '2px solid #4b286d', outline: 'none'
      }))
    }
  }

  const editableBlur = () => {
    const { textContent } = cellRef.current
    if (!textContent) {
      setDivStyle(p => ({
        ...p, background: 'transparent', borderBottom: '2px solid #d8d8d8', outline: 'none'
      }))
    }
  }
  const editableInput = (e) => {
    const { textContent } = e.currentTarget
    if (textContent && textContent.trim() !== '') {
      setDivStyle(p => ({
        ...p, background: 'white', borderBottom: '2px solid #4b286d', outline: 'none'
      }))
      addCommentHandler(e)
    } else {
      cellRef.current.innerHTML = ''
      setCellValue('')
      cellRef.current.blur()
      setDivStyle(p => ({
        ...p, background: 'transparent', borderBottom: '2px solid #d8d8d8', outline: 'none'
      }))
      addCommentHandler(e)
    }
  }

  return (
    <TextEditable
      className="editable"
      ref={cellRef}
      id={id}
      contentEditable={!disabled}
      style={divStyle}
      onClick={editableClick}
      onBlur={editableBlur}
      onInput={editableInput}
      content={cellValue && cellValue.length}
      placeholder={placeholder}
      maxLength="2000"
      dangerouslySetInnerHTML={{ __html: cellValue && cellValue.replace(/\n/g, '<br/>') }}
    />
  )
}
SelectInput.defaultProps = {
  placeholder: 'Type Here',
  disabled: false
}
SelectInput.propTypes = {
  id: PropTypes.string.isRequired,
  placeholder: PropTypes.string,
  addCommentHandler: PropTypes.func.isRequired,
  newCommentText: PropTypes.string.isRequired,
  disabled: PropTypes.bool
}
export default SelectInput
