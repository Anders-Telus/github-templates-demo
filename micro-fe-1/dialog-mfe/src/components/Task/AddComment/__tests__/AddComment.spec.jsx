import React from 'react'
import { mount } from 'enzyme'
import toJson from 'enzyme-to-json'
import AddComment from '../view'

describe('Custom AddComment component', () => {
  afterEach(() => {
    jest.clearAllMocks()
  })
  it('should match Snapshot', () => {
    const wrapper = mount(
      <AddComment
        id="add_new_comment"
        placeholder="Add Comment"
        addCommentHandler={jest.fn()}
        newCommentText=""
      />
    )
    expect(toJson(wrapper)).toMatchSnapshot()
  })
  it('Should active on click and backgroud will be white', () => {
    const wrapper = mount(
      <AddComment
        id="add_new_comment"
        placeholder="Add Comment"
        addCommentHandler={jest.fn()}
        newCommentText="Test New Comment"
      />
    )
    wrapper.find('div#add_new_comment').simulate('click')
    wrapper.find('div#add_new_comment').simulate('input', {
      target: {
        value: ''
      }
    })
    expect(wrapper.find('div#add_new_comment').props().style.background).toEqual('white')
  })
  it('Should background transparent on blur', () => {
    const wrapper = mount(
      <AddComment
        id="add_new_comment"
        placeholder="Add Comment"
        addCommentHandler={jest.fn()}
        newCommentText=""
      />
    )
    wrapper.find('div#add_new_comment').simulate('click')
    expect(wrapper.find('div#add_new_comment').props().style.background).toEqual('white')
    wrapper.find('div#add_new_comment').simulate('blur')
    expect(wrapper.find('div#add_new_comment').props().style.background).toEqual('transparent')
  })
  it('Should type in comment at event onInput', () => {
    const wrapper = mount(
      <AddComment
        id="add_new_comment"
        placeholder="Add Comment"
        addCommentHandler={jest.fn()}
        newCommentText=""
      />
    )
    wrapper.find('div#add_new_comment').simulate('click')
    wrapper.find('div#add_new_comment').simulate('input', {
      target: {
        value: ''
      }
    })
    expect(wrapper.find('div#add_new_comment').props().content).toEqual('')
  })
})
