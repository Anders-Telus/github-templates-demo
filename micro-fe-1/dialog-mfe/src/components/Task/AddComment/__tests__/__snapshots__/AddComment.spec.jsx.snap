// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Custom AddComment component should match Snapshot 1`] = `
<SelectInput
  addCommentHandler={[MockFunction]}
  disabled={false}
  id="add_new_comment"
  newCommentText=""
  placeholder="Add Comment"
>
  <styles__TextEditable
    className="editable"
    content=""
    contentEditable={true}
    dangerouslySetInnerHTML={
      Object {
        "__html": "",
      }
    }
    id="add_new_comment"
    maxLength="2000"
    onBlur={[Function]}
    onClick={[Function]}
    onInput={[Function]}
    placeholder="Add Comment"
    style={
      Object {
        "background": "transparent",
        "borderBottom": "2px solid #d8d8d8",
        "maxHeight": "132px",
        "minHeight": "30px",
        "overflowY": "scroll",
        "padding": "9px 8px 9px 8px",
        "width": "100%",
      }
    }
  >
    <div
      className="styles__TextEditable-sc-1nz0nez-0 fXdXMK editable"
      content=""
      contentEditable={true}
      dangerouslySetInnerHTML={
        Object {
          "__html": "",
        }
      }
      id="add_new_comment"
      maxLength="2000"
      onBlur={[Function]}
      onClick={[Function]}
      onInput={[Function]}
      placeholder="Add Comment"
      style={
        Object {
          "background": "transparent",
          "borderBottom": "2px solid #d8d8d8",
          "maxHeight": "132px",
          "minHeight": "30px",
          "overflowY": "scroll",
          "padding": "9px 8px 9px 8px",
          "width": "100%",
        }
      }
    />
  </styles__TextEditable>
</SelectInput>
`;
